js/ui/codingManager.js

```diff
@@ -0,0 +1,638 @@
+        // a window is slow to draw.
+        let firstFrameConnection = session.actorBuilder.connect('first-frame', Lang.bind(this, function() {
+            // reset the bouncing animation that was showed while Builder was loading
+            Tweener.removeTweens(session.buttonApp);
+            session.buttonApp.scale_y = 1;
+            session.buttonApp.scale_x = 1;
+            session.buttonApp.translation_y = 0;
+            session.buttonApp.translation_x = 0;
+
+            this._animate(session.actorApp, session.actorBuilder, Gtk.DirectionType.LEFT);
+
+            this._firstFrameConnections = this._firstFrameConnections.filter(function(conn) {
+                return conn != session.firstFrameConnection;
+            });
+            session.actorBuilder.disconnect(session.firstFrameConnection);
+            session.firstFrameConnection = null;
+
+            this._addButton(session);
+            session.buttonBuilder.show();
+            session.buttonApp.hide();
+
+            return false;
+        }));
+
+        this._prepareAnimate(session.actorApp, session.actorBuilder, Gtk.DirectionType.LEFT);
+        // Save the connection's id in the session and in a list too so we can
+        // get rid of it on kill-window-effects later */
+        session.firstFrameConnection = firstFrameConnection;
+        this._firstFrameConnections.push(firstFrameConnection);
+    },
+
+    _updateAppSizeAndPosition: function(window, session) {
+        let rect = session.actorApp.meta_window.get_frame_rect();
+        session.buttonApp.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.actorBuilder)
+            return;
+        session.actorBuilder.meta_window.move_resize_frame(false,
+                                                           rect.x,
+                                                           rect.y,
+                                                           rect.width,
+                                                           rect.height);
+    },
+
+    _updateBuilderSizeAndPosition: function(window, session) {
+        let rect = session.actorBuilder.meta_window.get_frame_rect();
+        session.buttonBuilder.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.actorApp)
+            return;
+        session.actorApp.meta_window.move_resize_frame(false,
+                                                       rect.x,
+                                                       rect.y,
+                                                       rect.width,
+                                                       rect.height);
+    },
+
+    _windowMinimized: function(shellwm, actor, session) {
+        // take the actor that we minimized and emit a minimized
+        // signal for it, though setting a flag internally so that
+        // we don't re-enter this signal handler.
```

Comment:
So what happens here is that when one of the windows (App or Builder) is minimized we do minimize the other one. So when doing this, this handler get called again for the window we just minimized. That is what the extra tracking is for.